### DISCUSSION TITLE
How to use the API Rules?

### DISCUSSION THREAD

**Discussion by paulrouge** [pocketbase/pocketbase#1157] [Q&A]
I'm trying to send a request to delete a post from a collection that has the following rule for the delete action:

`@request.auth.id = poster.id `    // "poster" is a relation-entry to "users"

When I try and call the 'delete' method from the client I get this error:

`"code":404,"message":"The requested resource wasn't found."`

I believe the `@request.auth.id` can not be found, should it be send with the header? If so, how?

The function:

<img width="662" alt="Screenshot 2022-12-01 at 17 43 38" src="https://user-images.githubusercontent.com/65511035/205110224-25454887-7a61-4216-9311-0d884810a19c.png">

_** Without the API rules set this does work._

Any ideas?

**Comment by ganigeorgiev** [comment]
The rule `@request.auth.id = poster.id` translates to: "Allow only authenticated user with id matching the poster.id to perform this operation".

So this means that you need to be authenticated before sending the delete request (eg. `pb.collection('users').authWithPassword('test@example.com', '123456')`).

The error is a generic 404 because the API rules can act also as filters and there is no easy way (without upfront parsing or static analysis) to know on which part of the filter expression the query will result to error/false.

**Reply by paulrouge** [reply]
Got it working! Thanks (also for the quick response)!

**Reply by ghost** [reply]
can you please tell me what this rule translate to "id = @request.auth.id"

**Reply by ganigeorgiev** [reply]
@JatinjayJAM The id of the current record must be equal to the current authenticated user id (this rule usually makes sense when used in auth collections).

**Reply by ghost** [reply]
thank you for your reply i am making a comment system where i want the user who commented to be able to update the comment what api rule should i use? will this rule work "Comment.user.id = @request.auth.id"

**Reply by ganigeorgiev** [reply]
@JatinjayJAM It is hard to say because it depends on your collections structure and what are the desired constraints you want to enforce, but if we assume that there is a `comments` collection with a "user" non-empty single `relation` field, then an example `comments` rules set could be:

```js
// list, view, delete
user = @request.auth.id

// create ("user = @request.auth.id" will also work for create)
@request.data.user = @request.auth.id

// update - roughly translates to: 
// allow changes of the record only if the current user field matches the auth.id
// AND
// don't allow users to change the comment's user value to another user id (aka. they can submit only the same user id or none)
user = @request.auth.id &&
(@request.data.user = @request.auth.id || @request.data.user:isset = false)
```

For more info about the API rules syntax, you can refer to https://pocketbase.io/docs/api-rules-and-filters/. If you need further help, please open a new discussion with more details about your collections structure and the desired behavior.

**Reply by hood25** [reply]
Can you please elaborate more, does that mean client have to input password in every update/delete request to satisfy the rule @request.auth.id?

**Reply by ganigeorgiev** [reply]
@hood25 A request is considered authenticated if it is send with valid `Authorization:TOKEN` header. The password is used only for the initial auth call to generate the token. If you have further questions, please consider opening a new discussion with more details about your use case.

**Comment by mwakaambrose** [comment]
This clarification about how the rules work has helped me too. Thank you @ganigeorgiev 💯
